{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sqlite3\n",
    "from Bio import SeqIO\n",
    "import random\n",
    "import re\n",
    "from jsonschema import validate\n",
    "\n",
    "\n",
    "class CodonDatabase:\n",
    "    def __init__(self,database=None,default_organism=None,rare_cutoff=None):\n",
    "        if database != None:\n",
    "            self.conn = sqlite3.connect(database)\n",
    "        else:\n",
    "            self.conn = sqlite3.connect(':memory:')\n",
    "        self.default_organism=default_organism\n",
    "        if rare_cutoff is None:\n",
    "            self.rare_cutoff = .1\n",
    "        else:\n",
    "            self.rare_cutoff=rare_cutoff\n",
    "        self.build_database()\n",
    "            \n",
    "    codon_table_validator = {\n",
    "          \"$schema\": \"http://json-schema.org/draft-07/schema#\",\n",
    "          #\"$id\": \"http://example.com/product.schema.json\",\n",
    "          \"title\": \"Codon Table Schema\",\n",
    "          \"description\": \"A JSON validator schema for Codon Tables\",\n",
    "          \"type\": \"object\",\n",
    "            \"properties\": {\n",
    "                \"id\": {\"description\": \"Unique identifier for the table. Usually a GenBank ID.\",\n",
    "                      \"type\": \"string\"},\n",
    "                \"organism\": {\"description\": \"Name of organism for the table. Human readable.\", \"type\": \"string\"},\n",
    "                \"description\": {\"description\": \"Description of the table or of the organism used to build the table. Human readable.\"},\n",
    "                \"transl_table\": {\"description\": \"Translation table that the codon table uses\", \"type\": \"integer\"},\n",
    "                \"codons\": {\"description\": \"Each individual codon in the table\", \"type\": \"array\", \"items\": {\n",
    "                    \"type\": \"object\",\n",
    "                    \"description\": \"A single codon in a table\",\n",
    "                    \"properties\": {\"codon\": {\"description\": \"Uppercase 3 letter non-degenerative DNA code\", \"pattern\": \"^[ATGC]*$\", \"type\": \"string\",\n",
    "                                            \"maxLength\":3,\"minLength\":3},\n",
    "                                  \"amino_acid\": {\"description\": \"The amino acid coded by the codon. Uppercase one latter code.\", \"pattern\":\"^[ARNDBCEQZGHILKMFPSTWYV*X]$\",\n",
    "                                                \"type\":\"string\", \"maxLength\":1,\"minLength\":1},\n",
    "                                  \"codon_count\": {\"description\": \"Count of codon occurrence in all genes in given organism or GenBank file.\", \"type\": \"integer\"}\n",
    "                                  }\n",
    "                }}\n",
    "            }\n",
    "        }\n",
    "\n",
    "    def build_database(self):\n",
    "        # Database schema of the Codon Database\n",
    "        CREATE_DB =\"\"\"\n",
    "            CREATE TABLE IF NOT EXISTS 'amino_acids' (\n",
    "                'amino_acid' TEXT PRIMARY KEY\n",
    "            );\n",
    "\n",
    "            CREATE TABLE IF NOT EXISTS 'organisms' (\n",
    "                'id' TEXT PRIMARY KEY,\n",
    "                'description' TEXT NOT NULL,\n",
    "                'organism' TEXT NOT NULL\n",
    "            );\n",
    "\n",
    "            CREATE TABLE IF NOT EXISTS 'transl_tables' (\n",
    "                'transl_table' INT PRIMARY KEY\n",
    "            );\n",
    "\n",
    "            CREATE TABLE IF NOT EXISTS 'codons' (\n",
    "                'codon' TEXT NOT NULL,\n",
    "                'codon_count' INTEGER NOT NULL DEFAULT 0,\n",
    "                'amino_acid' TEXT NOT NULL,\n",
    "                'transl_table' INTEGER NOT NULL,\n",
    "                'organism' TEXT NOT NULL,\n",
    "                FOREIGN KEY('amino_acid') REFERENCES 'amino_acids'('amino_acid'),\n",
    "                FOREIGN KEY('transl_table') REFERENCES 'transl_tables'('transl_table'),\n",
    "                FOREIGN KEY('organism') REFERENCES 'organisms'('id')\n",
    "                PRIMARY KEY('codon', 'amino_acid', 'organism')\n",
    "            );\n",
    "\n",
    "            \"\"\"\n",
    "        for x in CREATE_DB.split(';'):\n",
    "            self.conn.cursor().execute(x)\n",
    "\n",
    "        amino_acids = 'ARNDBCEQZGHILKMFPSTWYV*X'\n",
    "        self.conn.cursor().executemany('INSERT OR IGNORE INTO amino_acids(amino_acid) VALUES(?)', [(x,) for x in amino_acids])\n",
    "        self.conn.commit()\n",
    "\n",
    "        # transl_tables are for 3NF standardization\n",
    "\n",
    "    def choose_codon(self,aa,organism_id=None,banned=[],gc_bias=None,rare_cutoff=None):\n",
    "        if organism_id is None:\n",
    "            organism_id=self.default_organism\n",
    "        if rare_cutoff is None:\n",
    "            rare_cutoff=self.rare_cutoff\n",
    "\n",
    "        # Check that the amino acid is valid\n",
    "        valid_aa = 'ARNDBCEQZGHILKMFPSTWYV*'\n",
    "        if aa not in valid_aa:\n",
    "            return ValueError('{} not valid amino acid ({})'.format(aa,valid_aa))\n",
    "\n",
    "        # Check that all codons aren't banned\n",
    "        codons_per_aa = self.conn.cursor().execute('SELECT COUNT(*) FROM codons WHERE amino_acid = ? AND organism = ?', (aa,organism_id)).fetchone()[0]\n",
    "        if len(banned) >= codons_per_aa:\n",
    "            raise ValueError('Too many banned codons')\n",
    "\n",
    "        # 1. Build a table with a certain amino acid and a certain codon without any of the banned codons\n",
    "        # 2. Take cumulative probability and, using a random number input from python, select a codon\n",
    "        random_select = \"\"\"\n",
    "            WITH codons_selective AS ( WITH codon_gc AS\n",
    "            (\n",
    "                   SELECT codon, -- 2. Selects a table with desired amino acid, organism, and no codons from banned list. Also gives gc_content of each codon\n",
    "                          amino_acid,\n",
    "                          organism,\n",
    "                          codon_count,\n",
    "                          ( ( Length(Replace(( Replace(codon, 'T', '') ), 'A', '')) ) * 1.0 ) / Length( codon) AS gc_content\n",
    "                   FROM   (\n",
    "                                 SELECT * -- 1. Removes rare codons that occur fewer than 10% of the time, though this is configurable.\n",
    "                                 FROM   codons\n",
    "                                 WHERE  ((\n",
    "                                                      codon_count /\n",
    "                                                      (\n",
    "                                                               SELECT   Sum(codon_count)\n",
    "                                                               FROM     codons\n",
    "                                                               GROUP BY amino_acid)) > ?)) AS common_codons\n",
    "                   WHERE  amino_acid = ?\n",
    "                   AND    organism = ?\n",
    "                   AND    codon NOT IN {})\n",
    "            SELECT * -- 3. Selects for either high GC or low GC codons\n",
    "            FROM   codon_gc {})\n",
    "            SELECT   codon, -- Calculates cumulative probability of each codon based on occurrence, then chooses one, given a random number 0-1\n",
    "                     Cast(\n",
    "                            (\n",
    "                            SELECT Sum(codon_count)\n",
    "                            FROM   codons_selective AS c\n",
    "                            WHERE  c.codon <= codons_selective.codon) AS FLOAT) /\n",
    "                     (\n",
    "                            SELECT Sum(codon_count)\n",
    "                            FROM   codons_selective) AS probability\n",
    "            FROM     codons_selective\n",
    "            ORDER BY Abs(probability - Abs(Random() / 18446744073709551616 + 0.5)) -- Number comes from https://www.sqlite.org/lang_corefunc.html#random\n",
    "                     ASC limit 1\n",
    "            \"\"\" # https://stackoverflow.com/questions/50534961/sqlite-how-to-select-rows-based-on-probability-via-integer-value\n",
    "\n",
    "        if gc_bias == 'GC':\n",
    "            gc_bias_string = 'WHERE  gc_content = (SELECT Max(gc_content) FROM codon_gc)'\n",
    "        if gc_bias == 'AT':\n",
    "            gc_bias_string = 'WHERE  gc_content = (SELECT Min(gc_content) FROM codon_gc)'\n",
    "        else:\n",
    "            gc_bias_string = ''\n",
    "        try:\n",
    "            # Select a codon. First, add in a variable number of ? for banned codons, then run above statement. Fetch first codon of first result\n",
    "            r = self.conn.cursor().execute(random_select.format('(' + ','.join([\"?\" for _ in banned]) + ')',gc_bias_string), (rare_cutoff,aa,organism_id,)+tuple(banned)).fetchone()[0]\n",
    "        except Exception as e:\n",
    "            # If there is a value error, it is likely that the organism is not in the database\n",
    "            raise ValueError('Organism or amino acid not found in database')\n",
    "        return r\n",
    "\n",
    "    def optimize_sequence(self,protein_seq,organism_id):\n",
    "        if organism_id is None:\n",
    "            organism_id = self.default_organism\n",
    "        return ''.join([self.choose_codon(organism_id, codon) for codon in protein_seq])\n",
    "\n",
    "    def build_from_record(self,record):\n",
    "        c = self.conn.cursor()\n",
    "        c.execute('INSERT OR IGNORE INTO organisms(id,description,organism) VALUES(?,?,?)', (record.id, record.description,record.annotations['organism']))\n",
    "        self.default_organism = record.id\n",
    "        for feature in record.features:\n",
    "            if 'translation' in feature.qualifiers:\n",
    "                translation = feature.qualifiers['translation'][0] + '*'\n",
    "                seq = feature.extract(record).seq\n",
    "                if 'transl_table' in feature.qualifiers:\n",
    "                    transl_table = int(feature.qualifiers['transl_table'][0])\n",
    "                else:\n",
    "                    transl_table = 0\n",
    "                c.execute('INSERT OR IGNORE INTO transl_tables(transl_table) VALUES(?)', (transl_table,))\n",
    "                c.executemany('INSERT OR IGNORE INTO codons(amino_acid,codon,organism,transl_table) VALUES(?,?,?,?)', [(aa,str(seq[i*3:i*3+3]),record.id,transl_table) for i,aa in enumerate(translation)])\n",
    "                c.executemany('UPDATE codons SET codon_count = codon_count + 1 WHERE amino_acid = ? AND codon = ? AND organism = ?', [(aa,str(seq[i*3:i*3+3]),record.id) for i,aa in enumerate(translation)])\n",
    "        self.conn.commit()\n",
    "\n",
    "    def build_from_genbank(self,genbank_file):\n",
    "        for record in SeqIO.parse(genbank_file, \"genbank\"):\n",
    "            self.build_from_record(record)\n",
    "        return self\n",
    "\n",
    "    def codon_to_aa(self,codon,organism_id=None):\n",
    "        if organism_id is None:\n",
    "            organism_id = self.default_organism\n",
    "        return self.conn.cursor().execute('SELECT amino_acid FROM codons WHERE codon = ? AND organism = ? ORDER BY codon_count DESC', (codon,organism_id)).fetchone()[0]\n",
    "\n",
    "    def dump_codon_database(self):\n",
    "        return '\\n'.join([line for line in self.conn.iterdump()])\n",
    "\n",
    "    def available_codons(self,aa,banned,organism_id=None,rare_cutoff=None):\n",
    "        if organism_id is None:\n",
    "            organism_id = self.default_organism\n",
    "        if rare_cutoff is None:\n",
    "            rare_cutoff = self.rare_cutoff\n",
    "        return [{'codon':x[0],'weight':x[1]} for x in self.conn.cursor().execute(\"\"\"SELECT codon, 1.0*codon_count/(SELECT Sum(codon_count) FROM codons WHERE amino_acid = ?) as percent FROM codons WHERE amino_acid = ? AND organism = ? AND percent > ? AND codon NOT IN ({}) \"\"\".format('(' + ','.join([\"?\" for _ in banned]) + ')'), (aa,aa,organism_id,rare_cutoff)+tuple(banned)).fetchall()]\n",
    "\n",
    "    def export_table(self,organism_id=None):\n",
    "        if organism_id == None:\n",
    "            organism_id = self.default_organism\n",
    "            if organism_id == None:\n",
    "                raise ValueError('No organism_id given')\n",
    "        rows = self.conn.cursor().execute(\"SELECT o.id,o.description,o.organism,c.codon,c.codon_count,c.amino_acid,c.transl_table FROM organisms as o JOIN codons as c on c.organism=o.id WHERE id = ?\",(organism_id,)).fetchall()\n",
    "        organism_json = {'id':rows[0][0],'description':rows[0][1],'organism':rows[0][2],'transl_table':rows[0][6]}\n",
    "        organism_json['codons'] = [{'codon':row[3],'codon_count':row[4],'amino_acid':row[5]} for row in rows]\n",
    "        return organism_json\n",
    "    \n",
    "    def import_table(self,json_table):\n",
    "        validate(instance=json_table,schema=self.codon_table_validator)\n",
    "        transl_table = json_table['transl_table']\n",
    "        organism = json_table['id']\n",
    "        c = self.conn.cursor()\n",
    "        c.execute('INSERT OR IGNORE INTO organisms(id,description,organism) VALUES(?,?,?)',(organism,json_table['description'],json_table['organism']))\n",
    "        c.execute('INSERT OR IGNORE INTO transl_tables(transl_table) VALUES(?)', (transl_table,))\n",
    "        c.executemany('INSERT OR IGNORE INTO codons(amino_acid,codon,codon_count,organism,transl_table) VALUES(?,?,?,?,?)', [(codon['amino_acid'],codon['codon'],codon['codon_count'],organism,transl_table) for codon in json_table['codons']])\n",
    "        self.conn.commit()\n",
    "    \n",
    "class Triplet:\n",
    "    def __init__(self,table,codon,organism_id=None,rare_cutoff=None):\n",
    "        self.table = table\n",
    "        self.codon = codon.upper()\n",
    "        self.last_codon = self.codon\n",
    "        self.aa = self.table.codon_to_aa(codon,organism_id)\n",
    "        self.organism_id = organism_id\n",
    "        self.banned = [codon]\n",
    "        self.rare_cutoff = rare_cutoff\n",
    "    \n",
    "    def change_codon(self, gc_bias=None):\n",
    "        new_codon = self.table.choose_codon(self.aa,self.organism_id,self.banned,gc_bias=gc_bias)\n",
    "        self.last_codon = self.codon\n",
    "        self.codon = new_codon\n",
    "        self.banned.append(new_codon)\n",
    "        return new_codon\n",
    "    \n",
    "    def available_codons(self):\n",
    "        return self.table.available_codons(self.aa,self.banned,self.organism_id,self.rare_cutoff)\n",
    "        \n",
    "    def __str__(self):\n",
    "        return self.codon\n",
    "\n",
    "\n",
    "class CodingSequence:\n",
    "    def __init__(self,table,sequence,organism_id=None):\n",
    "        # Add in verification\n",
    "        self.triplets = [Triplet(table,x,organism_id) for x in re.findall('...',sequence)]\n",
    "        \n",
    "    def __str__(self):\n",
    "        return ''.join([str(x) for x in self.triplets])\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "c = CodonDatabase('example.db','NC_000964.3')\n",
    "#c.build_from_genbank('sequence.gb')\n",
    "k1_5 = 'atgcaaggcctacacgctattcaacttcaacttgaagaagaaatgtttaacggcggtatccgtcgctttgaagcggaccaacaacgccagattgcatccggtaatgaatcagacacggcatggaatcgccgcttattgtccgagttaatcgcgccaatggctgaaggtattcaggcatacaaggaagagtatgaaggtaaaagaggccgtgcaccgcgtgcattagctttcattaactgcgtagaaaacgaagtggcagcatatatcacgatgaaaatcgttatggatatgctgaacacggatgtaaccttgcaggctatagccatgaatgtagctgaccgcattgaggaccaagtacgttttagcaagctggaaggtcacgccgccaaatactttgaaaaagttaagaagtcacttaaggcaagtaagactaaatcatatcgccatgcgcacaacgtagcggtagtggctgagaagtcagtagctgaccgtgacgctgatttctcccgctgggaggcatggcctaaagacaccttgctgcaaattgggatgaccttgcttgaaatcttagagaatagcgtattcttcaacgggcaacctgtcttcctccgcaccttgcgcactaatggcggcaaacatggtgtttactacctacagactagtgaacacgtaggtgagtggataactgcattcaaagagcacgtagcgcaactgagtcctgcctatgctccttgcgtcatccctccgcgtccgtgggtatcaccttttaacggcggtttccacactgagaaagtagcaagccgtattcgtctggtaaaaggaaaccgcgaacacgtccgcaagctgaccaaaaagcaaatgccagaggtttacaaggctgttaacgcgttgcaggcgactaaatggcaggttaacaaggaagttttacaggttgtggaagacgtcatccgtctagacctaggttatggtgtaccttcctttaaaccactcattgaccgcgagaacaagccagctaatccagtgccgctagaatttcagcacctacggggccgtgaactgaaagaaatgcttacgccggaacaatggcaagcctttatcaactggaaaggtgaatgtactaagctgtacaccgctgaaactaagcgcggaagcaaatcggcggcaaccgttcgcatggttggtcaggcccgtaaatatagccagttcgacgcaatctacttcgtgtatgcactggacagccgcagccgcgtctacgcgcaatctagcacactctcaccgcaatcaaatgacttgggcaaggccttgctccgttttaccgaagggcagcgtcttgatagcgctgaggcgcttaagtggtttttggtgaacggggctaataactggggttgggataagaaaacttttgacgtgcgcaccgctaacgtgctggatagtgaatttcaagacatgtgccgcgacattgcagcggatccgctgaccttcactcaatgggtaaatgccgactccccttacggcttccttgcatggtgctttgaatatgcgcgttatctggatgcactggatgaaggcacgcaagaccaattcatgacgcacctcccagtccatcaagatggtagttgttctggtatccagcactacagtgctatgctacgcgatgcagtaggtgcgaaagcagtaaaccttaagccctctgactctcctcaagatatttatggtgccgttgcgcaggtagtaattcagaagaattatgcatacatgaatgcagaggatgcggaaaccttcacttctggcagcgtgactttaacaggtgcggagctgcgtagtatggctagtgcgtgggatatgataggaatcactcgcggcctgaccaaaaagcccgtaatgacactaccttatggcagcacacgtctaacctgccgtgagtcagtgattgattatatcgttgatttagaagaaaaagaggcccaacgggctattgcggaagggcgtaccgccaatcctgtacacccttttgataatgaccgtaaagacagcctgacacctagcgcagcttataactatatgacagctttaatctggccttctatttcggaagtggttaaagcccctatagtggcaatgaaaatgattcgtcagcttgcccgtttcgcagctaaaaggaatgaaggcttagagtataccctgcctactggcttcatcttgcaacaaaagattatggctactgatatgctccgcgtatctacttgcttgatgggagaaatcaagatgagtctacagattgaaacagacgtagtggatgaaacggcaatgatgggcgctgctgctcctaactttgtgcatggtcatgatgccagccaccttatcttaacagtctgcgaccttgttgataaagggattacatctatcgcagttattcatgactcttttggcactcatgcaggccgtacagccgaccttcgtgatagcttaagggcagaaatggtgaagatgtatcaaggccgtaatgcactgcaaagcctgctagatgagcacgaagaacgctggttagttgataccggaatacaagtaccagagcaaggggagtttgaccttaacgaaatcttagtttcagactattgcttcgcataa'.upper()\n",
    "s = CodingSequence(c,k1_5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
